pkg/util/alibaba/alibaba_test.go

```diff
@@ -30,3 +32,24 @@ func TestGetHostname(t *testing.T) {
 	assert.Equal(t, expected, val)
 	assert.Equal(t, lastRequest.URL.Path, "/latest/meta-data/instance-id")
 }
+
+func TestGetNTPHosts(t *testing.T) {
+	expectedHosts := []string{
+		"ntp.cloud.aliyuncs.com", "ntp1.cloud.aliyuncs.com", "ntp2.cloud.aliyuncs.com", "ntp3.cloud.aliyuncs.com",
+		"ntp4.cloud.aliyuncs.com", "ntp5.cloud.aliyuncs.com", "ntp6.cloud.aliyuncs.com", "ntp7.cloud.aliyuncs.com",
+		"ntp8.cloud.aliyuncs.com", "ntp9.cloud.aliyuncs.com", "ntp10.cloud.aliyuncs.com", "ntp11.cloud.aliyuncs.com",
+		"ntp12.cloud.aliyuncs.com",
```

Comment:
are all of these NTP servers reachable from any VM on Alibaba's cloud, or are there restrictions?

The current logic of the ntp check would simply pick one of the hosts at random, so assumes that all the hosts are reachable

Reply:
The restrictions are here: https://www.alibabacloud.com/help/doc-detail/92704.htm
I wasn't sure which to pick.

As for reachability from the VMs, all of them are internal and should be reachable. For Azure, the docs are unclear on how to access the private NTP server or if there is one, so that one is the public windows server (not sure if it gets routed internally to a private one).

I tested it on all but Tencent, which we do not have a sandbox for.

Reply:
thanks for clarifying. About the various ntp servers on Alibaba, are all of them reachable whatever the network type (classic or VPC)? The table seems to suggest otherwise, but it's worth giving it a try.

Reply:
please update this comment thread with the results of your testing :)

Reply:
For Alibaba NTP servers, there are three types of NTP servers:

* Private ones in what they call "classic network"
* Private ones in a regular VPC
* Public ones

As we do not know cross-accessibility with the classic network and VPC, we will use the Public NTP servers. They should not be less accessible than pool.ntp.org ones we use currently.